subsys/net/lib/prometheus/collector.c

```diff
@@ -0,0 +1,145 @@
+/*
+ * Copyright (c) 2024 Mustafa Abdullah Kus, Sparse Technology
+ *
+ * SPDX-License-Identifier: Apache-2.0
+ */
+
+#include <zephyr/net/prometheus/collector.h>
+
+#include <zephyr/net/prometheus/metric.h>
+#include <zephyr/net/prometheus/histogram.h>
+#include <zephyr/net/prometheus/summary.h>
+#include <zephyr/net/prometheus/counter.h>
+#include <zephyr/net/prometheus/gauge.h>
+
+#include <stdlib.h>
+#include <string.h>
+
+#include <zephyr/kernel.h>
+#include <zephyr/sys/iterable_sections.h>
+
+#include <zephyr/logging/log.h>
+LOG_MODULE_REGISTER(pm_collector, CONFIG_PROMETHEUS_LOG_LEVEL);
+
+int prometheus_collector_register_metric(struct prometheus_collector *collector,
+					 struct prometheus_metric *metric)
+{
+	if (!collector || !metric) {
+		LOG_ERR("Invalid arguments");
+		return -EINVAL;
+	}
+
+	LOG_DBG("Registering metric type=%d", metric->type);
+
+	for (int i = 0; i < CONFIG_PROMETHEUS_MAX_METRICS; i++) {
+		if (!collector->metric[i]) {
+			collector->metric[i] = metric;
+			collector->size++;
+			return 0;
+		}
+	}
+
+	return -ENOMEM;
+}
+
+const struct prometheus_counter *prometheus_get_counter_metric(const char *name)
+{
+	STRUCT_SECTION_FOREACH(prometheus_counter, entry) {
+		LOG_DBG("entry->name: %s", entry->base->name);
+
+		if (strncmp(entry->base->name, name, sizeof(entry->base->name)) == 0) {
+			LOG_DBG("Counter found %s=%s", entry->base->name, name);
+			return entry;
+		}
+	}
+
+	LOG_DBG("Counter not found %s", name);
+
+	return NULL;
+}
+
+const struct prometheus_gauge *prometheus_get_gauge_metric(const char *name)
+{
+	STRUCT_SECTION_FOREACH(prometheus_gauge, entry) {
+		LOG_DBG("entry->name: %s", entry->base->name);
+
+		if (strncmp(entry->base->name, name, sizeof(entry->base->name)) == 0) {
+			LOG_DBG("Counter found %s=%s", entry->base->name, name);
+			return entry;
+		}
+	}
+
+	LOG_DBG("Counter not found %s", name);
+
+	return NULL;
+}
+
+const struct prometheus_histogram *prometheus_get_histogram_metric(const char *name)
+{
+	STRUCT_SECTION_FOREACH(prometheus_histogram, entry) {
+		LOG_DBG("entry->name: %s", entry->base->name);
+
+		if (strncmp(entry->base->name, name, sizeof(entry->base->name)) == 0) {
+			LOG_DBG("Counter found %s=%s", entry->base->name, name);
+			return entry;
+		}
+	}
+
+	LOG_DBG("Counter not found %s", name);
```

Comment:
And here `Histogram not found`?

Suggestion for the these error debug prints
```
LOG_DBG("%s %s not found", "Histogram", name);
```
And just change the type of the item not found. This way the base string is shared between different functions and no new copy is stored in flash.

Reply:
Good point. Missed out. Will fix and push right now.